core/src/main/scala/org/ensime/core/debug/VirtualMachineManager.scala

```diff
@@ -0,0 +1,194 @@
+    startFunc: (ScalaVirtualMachine) => Unit = _ => {}
+  ): Unit = synchronized {
+    import scala.concurrent.duration._
+
+    // Start the specific debugger based on mode and retrieve a
+    // ScalaVirtualMachine instance from launching or attaching
+    val (d, s) = mode match {
+      case VmStart(commandLine) =>
+        val d = LaunchingDebugger(
+          className = commandLine,
+          jvmOptions = options,
+          suspend = true
+        ).withPending(dummyScalaVirtualMachine)
+
+        val s = d.start(timeout = 10.seconds, startProcessingEvents = false)
+
+        (d, s)
+
+      case VmAttach(hostname, port) =>
+        val d = AttachingDebugger(
+          port = port.toInt,
+          hostname = hostname
+        ).withPending(dummyScalaVirtualMachine)
+
+        val s = d.start(timeout = 10.seconds, startProcessingEvents = false)
+
+        (d, s)
+    }
+
+    // Bind our event handles like class prepare and thread start/stop
+    bindEvents(s)
+
+    // Mark our debugger and acquired virtual machine as ready
+    debugger = Some(d)
+    vm = Some(s)
+    this.mode = Some(mode)
+
+    // Invoke our start function, ignoring any error that may arise
+    withVM(globalStartFunc)
+    withVM(startFunc)
+
+    // Begin processing events
+    s.startProcessingEvents()
+
+    (d, s)
+  }
+
+  /**
+   * Stops and shuts down the currently-active debugger and its associated
+   * virtual machines.
+   *
+   * @param stopFunc Optional function to call before the virtual machine
+   *                 shuts down
+   */
+  def stop(stopFunc: (ScalaVirtualMachine) => Unit = _ => {}): Unit = synchronized {
+    // Invoke our stop function, ignoring any error that may arise
+    if (hasActiveVM) {
+      withVM(globalStopFunc)
+      withVM(stopFunc)
+    }
+
+    clear()
+  }
+
+  /** Clears the active mode, vm, and debugger. */
+  private def clear(): Unit = synchronized {
+    // Clear our associated mode
+    mode = None
+
+    // Dispose of the virtual machine and discard the reference
+    Try(vm.foreach(_.underlyingVirtualMachine.dispose()))
+    vm = None
+
+    // Shutdown the associated debugger
+    Try(debugger.foreach(_.stop()))
+    debugger = None
+  }
+
+  /**
+   * Returns whether or not this manager has an active JVM that is being
+   * debugged.
+   *
+   * @return True if a remote virtual machine is connected, otherwise false
+   */
+  def hasActiveVM: Boolean = vm.nonEmpty
+
+  /**
+   * Retrieves the active JVM if available and runs the specified action on
+   * top of it.
+   *
+   * @param action The action to evaluate on top of the JVM
+   * @tparam T The expected return value from the action
+   * @return Some containing the result if successful, otherwise None
+   */
+  def withVM[T](action: (ScalaVirtualMachine => T)): Try[T] = synchronized {
+    if (!hasActiveVM) {
+      val error = new NoActiveVirtualMachineException
+      log.error("No VM active for debugging!", error)
+      return Failure(error)
```

Comment:
if you're wanting a `Try` interface here, I don't think it makes sense to define a new domain exception for it. Just use an existing exception type, such as `IllegalStateException` with an appropriate message.